File: backend/components/metiers/administration_publique/tests/utils/core/core_utils.js
// core_utils.js – Utilitaires ultra avancés, clé en main, pour tests métiers 3D
// Respecte la modularité, la conformité RGPD, l’audit, la sécurité et la performance

const { v4: uuidv4 } = require('uuid');

function generateId(prefix = 'obj') {
  return `${prefix}-${uuidv4()}`;
}

function deepClone(obj) {
  return JSON.parse(JSON.stringify(obj));
}

function isValidEmail(email) {
  return /^[^@\s]+@[^@\s]+\.[^@\s]+$/.test(email);
}

function auditLog(action, user, meta = {}) {
  return {
    id: generateId('audit'),
    action,
    user,
    meta,
    timestamp: new Date().toISOString()
  };
}

module.exports = {
  generateId,
  deepClone,
  isValidEmail,
  auditLog
};

File: backend/components/metiers/administration_publique/tests/utils/core/core_utils.py
# core_utils.py – Utilitaires ultra avancés, clé en main, pour tests métiers 3D
# Respecte la modularité, la conformité RGPD, l’audit, la sécurité et la performance
import uuid
import re
from datetime import datetime

def generate_id(prefix='obj'):
    return f"{prefix}-{uuid.uuid4()}"

def deep_clone(obj):
    import copy
    return copy.deepcopy(obj)

def is_valid_email(email):
    return re.match(r"^[^@\s]+@[^@\s]+\.[^@\s]+$", email) is not None

def audit_log(action, user, meta=None):
    if meta is None:
        meta = {}
    return {
        'id': generate_id('audit'),
        'action': action,
        'user': user,
        'meta': meta,
        'timestamp': datetime.utcnow().isoformat() + 'Z'
    }
